### PR TITLE
Add priority to defaultOn plugins list

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**What this PR does / why we need it**:
podPriority plugin needs to be on by default since priority graduated to beta and is on by default in features
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/cc @liggitt @bsalamat 
**Release note**:

```release-note
kube-apiserver: the `Priority` admission plugin is now enabled by default when using `--enable-admission-plugins`. If using `--admission-control` to fully specify the set of admission plugins, the `Priority` admission plugin should be added if using the `PodPriority` feature, which is enabled by default in 1.11.
```

### CODE REVIEW

File: pkg/kubeapiserver/options/plugins.go [hunk 1]
```diff
@@ -137,6 +137,7 @@ func DefaultOffAdmissionPlugins() sets.String {
 		mutatingwebhook.PluginName,          //MutatingAdmissionWebhook
 		validatingwebhook.PluginName,        //ValidatingAdmissionWebhook
 		resourcequota.PluginName,            //ResourceQuota
+		podpriority.PluginName,              //PodPriority
```
Comment (liggitt) [inline pkg/kubeapiserver/options/plugins.go]: only add this if the PodPriority feature gate is on?

Comment (ravisantoshgudimetla) [reply pkg/kubeapiserver/options/plugins.go]: You mean, import kube_features and check if podPriority is enabled and then append podpriority.PluginName to list?

Comment (liggitt) [reply pkg/kubeapiserver/options/plugins.go]: yes

Comment (bsalamat) [reply pkg/kubeapiserver/options/plugins.go]: I am confused! Isn't this the `DefaultOff` list?!

Comment (liggitt) [reply pkg/kubeapiserver/options/plugins.go]: this is computing DefaultOff by taking all registered plugins and subtracting the defaultOn list

Comment (bsalamat) [reply pkg/kubeapiserver/options/plugins.go]: Ok, nevermind. The function name confused me.

**Review by bsalamat — APPROVED** [review]
/lgtm

**Comment by bsalamat** [comment]
Please see the tests failures. Some of the failures may be relevant.

File: pkg/kubeapiserver/options/plugins.go (hunk 1 above)
Comment (bsalamat) [reply pkg/kubeapiserver/options/plugins.go]: > this is computing DefaultOff by taking all registered plugins and subtracting the defaultOn list

Yes, I actually figured before seeing your comment. It is a very odd combination that we build the defaultOn list in the defaultOff function!

**Comment by liggitt** [comment]
test failure is because we are saving and re-posting pods to the API:
https://github.com/kubernetes/kubernetes/blob/f7d95e208ac4d39e0e7a625d95dbc5ee8d6a1bfa/hack/make-rules/test-cmd-util.sh#L363-L364
https://github.com/kubernetes/kubernetes/blob/f7d95e208ac4d39e0e7a625d95dbc5ee8d6a1bfa/hack/make-rules/test-cmd-util.sh#L396-L399

this admission plugin sets an attribute that is required to be cleared before an exported pod can be recreated. that would have been good to know/consider before promoting.

our options are to ignore/override the user-specified priority on creation, or continue rejecting exported/recreated pods and requiring users to manually clear priorities.

**Comment by bsalamat** [comment]
> our options are to ignore/override the user-specified priority on creation, or continue rejecting exported/recreated pods and requiring users to manually clear priorities.

I would vote for rejecting and requiring users to manually clear priorities. This would help avoid/fix unintended errors in pod spec.

**Comment by ravisantoshgudimetla** [comment]
> that makes it more questionable to enable this admission plugin by default. it means that workflows that previously worked seamlessly will have to be adjusted, even if the originally posted pod specs did not do anything with pod priority or priority class

Do people export and recreate pods, is this normal workflow?

> How about letting the admission controller resolve the value of priority again on pod updates if the priority value is not provided? Will it fix the problem?

This might solve the problem but we need to document this behaviour.

**Comment by liggitt** [comment]
> I would vote for rejecting and requiring users to manually clear priorities. This would help avoid/fix unintended errors in pod spec.

an alternative would be to error if the computed priority differs from an explicit priority set in the pod spec. this would avoid confusion where the explicit priority was ignored, but also avoid errors in the following cases:
* an export/create of a pod on the same cluster
* an export/create of a pod between clusters sharing identical priority class definitions
* an export/create of a pod between clusters that are not making use of priorities (so all pods get `priority: 0`)

a fix for the other test failure (`kubectl create -f pod.yaml && kubectl replace -f pod.yaml`) could be to have the admission plugin preserve existing priority on update if the incoming update has no priority set.

I opened a PR at https://github.com/kubernetes/kubernetes/pull/65739 with those two additions on top of this PR. Since we effectively have to enable this admission plugin by default to preserve the integrity of the priority API, I think the burden is on us to make sure it is as non-disruptive as possible.

**Comment by ravisantoshgudimetla** [comment]
/close

Closing this in favour of #65739.

Thanks @liggitt

**Pull request closed without merging**